File: app/src/main/java/com/muratcan/apps/petvaccinetracker/viewmodel/PetViewModel.java
package com.muratcan.apps.petvaccinetracker.viewmodel;

import android.app.Application;
import android.os.Handler;
import android.os.Looper;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import com.muratcan.apps.petvaccinetracker.model.Pet;
import com.muratcan.apps.petvaccinetracker.model.Vaccine;
import com.muratcan.apps.petvaccinetracker.repository.PetRepository;
import com.muratcan.apps.petvaccinetracker.util.FirebaseHelper;

import java.util.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.stream.Collectors;

public class PetViewModel extends AndroidViewModel {
    private final PetRepository repository;
    private final FirebaseHelper firebaseHelper;
    private final MutableLiveData<Boolean> success = new MutableLiveData<>();
    private final MutableLiveData<List<Vaccine>> vaccines = new MutableLiveData<>();
    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);
    private final MutableLiveData<String> error = new MutableLiveData<>();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final MutableLiveData<Pet> currentPet = new MutableLiveData<>();
    private LiveData<List<Vaccine>> vaccinesLiveData;
    private final LiveData<List<Pet>> allPetsLiveData;
    private final MutableLiveData<String> searchQuery = new MutableLiveData<>("");
    private final MutableLiveData<String> sortOrder = new MutableLiveData<>("none");
    private final LiveData<List<Pet>> filteredAndSortedPets;

    public PetViewModel(Application application) {
        super(application);
        repository = new PetRepository(application);
        firebaseHelper = new FirebaseHelper();
        allPetsLiveData = repository.getAllPets();
        
        // Transform allPetsLiveData based on search query and sort order
        filteredAndSortedPets = Transformations.switchMap(searchQuery, query -> 
            Transformations.switchMap(sortOrder, order ->
                Transformations.map(allPetsLiveData, pets -> {
                    List<Pet> filteredList;
                    if (query == null || query.isEmpty()) {
                        filteredList = new ArrayList<>(pets);
                    } else {
                        filteredList = pets.stream()
                            .filter(pet -> pet.getName().toLowerCase().contains(query.toLowerCase()) ||
                                    (pet.getBreed() != null && pet.getBreed().toLowerCase().contains(query.toLowerCase())) ||
                                    pet.getType().toLowerCase().contains(query.toLowerCase()))
                            .collect(Collectors.toList());
                    }

                    // Apply sorting
                    switch (order) {
                        case "name":
                            Collections.sort(filteredList, (p1, p2) -> 
                                p1.getName().compareToIgnoreCase(p2.getName()));
                            break;
                        case "date":
                            Collections.sort(filteredList, (p1, p2) -> {
                                if (p1.getBirthDate() == null) return 1;
                                if (p2.getBirthDate() == null) return -1;
                                return p2.getBirthDate().compareTo(p1.getBirthDate());
                            });
                            break;
                    }
                    return filteredList;
                })
            )
        );
    }

    public LiveData<Boolean> getSuccess() {
        return success;
    }

    public LiveData<List<Vaccine>> getVaccines() {
        return vaccinesLiveData != null ? vaccinesLiveData : vaccines;
    }

    public LiveData<List<Pet>> getPets() {
        return filteredAndSortedPets;
    }

    public LiveData<Boolean> getIsLoading() {
        return isLoading;
    }

    public LiveData<String> getError() {
        return error;
    }

    public void loadPets() {
        isLoading.setValue(true);
        // The pets are automatically loaded through LiveData from Room
        isLoading.setValue(false);
    }

    public void sortByName() {
        sortOrder.setValue("name");
    }

    public void sortByDate() {
        sortOrder.setValue("date");
    }

    public void clearSort() {
        sortOrder.setValue("none");
    }

    public void filterPets(String query) {
        searchQuery.setValue(query);
    }

    public void loadVaccines(long petId) {
        clearError();
        setLoading(true);
        // Use LiveData for continuous updates
        vaccinesLiveData = repository.getVaccinesForPetLive(petId);
        setLoading(false);
    }

    public void refreshVaccines(long petId) {
        clearError();
        setLoading(true);
        // Use callback version for immediate refresh
        repository.getVaccinesForPet(petId,
            vaccineList -> mainHandler.post(() -> {
                vaccines.setValue(vaccineList);
                setLoading(false);
                clearError();
            }),
            errorMessage -> mainHandler.post(() -> {
                setError(errorMessage);
                setLoading(false);
            })
        );
    }

    public void addPet(Pet pet) {
        clearError(); // Clear any previous errors
        setLoading(true);
        repository.addPet(pet,
            () -> mainHandler.post(() -> {
                success.setValue(true);
                setLoading(false);
                clearError(); // Clear any previous errors on success
            }),
            errorMessage -> mainHandler.post(() -> {
                setError(errorMessage);
                setLoading(false);
            })
        );
    }

    public LiveData<Pet> getCurrentPet() {
        return currentPet;
    }

    public void loadPetById(long petId) {
        clearError();
        setLoading(true);
        repository.getPetById(petId,
            pet -> mainHandler.post(() -> {
                currentPet.setValue(pet);
                setLoading(false);
                clearError();
            }),
            errorMessage -> mainHandler.post(() -> {
                setError(errorMessage);
                setLoading(false);
            })
        );
    }

    private void setLoading(boolean loading) {
        mainHandler.post(() -> isLoading.setValue(loading));
    }

    private void setError(String errorMessage) {
        mainHandler.post(() -> error.setValue(errorMessage));
    }

    private void clearError() {
        mainHandler.post(() -> error.setValue(null));
    }
} 

File: app/src/main/java/com/muratcan/apps/petvaccinetracker/viewmodel/PetViewModel.kt
package com.muratcan.apps.petvaccinetracker.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.asLiveData
import androidx.lifecycle.viewModelScope
import com.muratcan.apps.petvaccinetracker.model.Pet
import com.muratcan.apps.petvaccinetracker.model.Vaccine
import com.muratcan.apps.petvaccinetracker.model.VaccineHistoryItem
import com.muratcan.apps.petvaccinetracker.repository.PetRepository
import com.muratcan.apps.petvaccinetracker.util.FirebaseHelper
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import timber.log.Timber

class PetViewModel(application: Application) : AndroidViewModel(application) {
    private val repository: PetRepository
    private val firebaseHelper: FirebaseHelper = FirebaseHelper()

    private val _pets = MutableStateFlow<List<Pet>>(emptyList())
    val pets: StateFlow<List<Pet>> = _pets.asStateFlow()
    private val _petsLiveData: LiveData<List<Pet>> = pets.asLiveData()

    private val _currentPet = MutableStateFlow<Pet?>(null)
    val currentPet: StateFlow<Pet?> = _currentPet.asStateFlow()
    private val _currentPetLiveData: LiveData<Pet?> = currentPet.asLiveData()

    private val _vaccines = MutableStateFlow<List<Vaccine>>(emptyList())
    val vaccines: StateFlow<List<Vaccine>> = _vaccines.asStateFlow()
    private val _vaccinesLiveData: LiveData<List<Vaccine>> = vaccines.asLiveData()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    private val _isLoadingLiveData: LiveData<Boolean> = isLoading.asLiveData()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()
    private val _errorLiveData: LiveData<String?> = error.asLiveData()

    private val _searchQuery = MutableStateFlow("")
    private val _sortOrder = MutableStateFlow("none")

    init {
        repository = PetRepository(application.applicationContext)
        loadPets()
        observePets()
    }

    private fun observePets() {
        viewModelScope.launch {
            try {
                val currentUserId = firebaseHelper.getCurrentUserId()
                if (currentUserId == null) {
                    _error.value = "User not logged in"
                    _isLoading.value = false
                    return@launch
                }
                
                repository.getPetsForUser(currentUserId).collect { petsList ->
                    val filteredAndSorted = applyFiltersAndSort(petsList)
                    _pets.value = filteredAndSorted
                    _isLoading.value = false
                }
            } catch (e: Exception) {
                _error.value = e.message
                _isLoading.value = false
            }
        }
    }

    private fun applyFiltersAndSort(pets: List<Pet>): List<Pet> {
        var result = if (_searchQuery.value.isEmpty()) {
            pets
        } else {
            pets.filter { pet ->
                pet.name.contains(_searchQuery.value, ignoreCase = true) ||
                pet.type.contains(_searchQuery.value, ignoreCase = true) ||
                pet.breed.contains(_searchQuery.value, ignoreCase = true)
            }
        }

        when (_sortOrder.value) {
            "name" -> result = result.sortedBy { it.name.lowercase() }
            "date" -> result = result.sortedByDescending { it.birthDate }
            else -> { /* no sorting */ }
        }

        Timber.d("Sorting by: ${_sortOrder.value}, Results: ${result.size} pets")
        return result
    }

    fun loadPets() {
        refreshPets()
    }

    fun refreshPets() {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                val currentUserId = firebaseHelper.getCurrentUserId()
                if (currentUserId == null) {
                    _error.value = "User not logged in"
                    _isLoading.value = false
                    return@launch
                }
                
                // Get first emission only
                val petsList = repository.getPetsForUser(currentUserId).first()
                val filteredAndSorted = applyFiltersAndSort(petsList)
                _pets.value = filteredAndSorted
                _isLoading.value = false
            } catch (e: Exception) {
                _error.value = e.message
                _isLoading.value = false
            }
        }
    }

    fun loadPetById(petId: Long) {
        viewModelScope.launch {
            try {
                val pet = repository.getPetByIdFlow(petId)
                _currentPet.value = pet
            } catch (e: Exception) {
                _error.value = e.message
            }
        }
    }

    fun loadVaccines(petId: Long) {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                repository.getVaccinesForPetFlow(petId).collect { vaccinesList ->
                    _vaccines.value = vaccinesList
                    _isLoading.value = false
                }
            } catch (e: Exception) {
                _error.value = e.message
                _isLoading.value = false
            }
        }
    }

    // Alias for Java code
    fun refreshVaccines(petId: Long) = loadVaccines(petId)

    fun addPet(pet: Pet) {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                Timber.d("Starting to add pet: ${pet.name}")
                
                // Get current user ID
                val currentUserId = firebaseHelper.getCurrentUserId()
                if (currentUserId == null) {
                    throw IllegalArgumentException("User not logged in")
                }
                pet.userId = currentUserId
                
                // Validate all required fields
                if (pet.name.isNullOrBlank()) {
                    throw IllegalArgumentException("Pet name cannot be null or empty")
                }
                if (pet.type.isNullOrBlank()) {
                    throw IllegalArgumentException("Pet type cannot be null or empty")
                }
                if (pet.breed.isNullOrBlank()) {
                    pet.breed = "Unknown" // Set a default value for breed if null
                }
                if (pet.birthDate == null) {
                    throw IllegalArgumentException("Birth date cannot be null")
                }
                
                Timber.d("Validated pet details - Name: ${pet.name}, Type: ${pet.type}, Breed: ${pet.breed}, UserId: ${pet.userId}, BirthDate: ${pet.birthDate}")
                
                // Add pet and get its ID
                val id = repository.addPet(pet)
                
                if (id > 0) {
                    Timber.d("Pet added successfully with ID: $id")
                    pet.id = id
                    
                    // Force a refresh to update the UI
                    val petsList = repository.getPetsForUser(currentUserId).first()
                    val filteredAndSorted = applyFiltersAndSort(petsList)
                    _pets.value = filteredAndSorted
                } else {
                    Timber.e("Failed to add pet - returned ID was $id")
                    _error.value = "Failed to add pet to database"
                }
            } catch (e: Exception) {
                Timber.e(e, "Error adding pet: ${e.message}")
                _error.value = "Error adding pet: ${e.message}"
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun updatePet(pet: Pet) {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                repository.updatePet(pet)
                _isLoading.value = false
            } catch (e: Exception) {
                _error.value = e.message
                _isLoading.value = false
            }
        }
    }

    fun deletePet(pet: Pet) {
        viewModelScope.launch {
            try {
                repository.deletePet(pet)
                _error.value = null
            } catch (e: Exception) {
                _error.value = e.message
            }
        }
    }

    fun filterPets(query: String) {
        _searchQuery.value = query
        viewModelScope.launch {
            try {
                val currentUserId = firebaseHelper.getCurrentUserId()
                if (currentUserId == null) {
                    _error.value = "User not logged in"
                    return@launch
                }
                
                val petsList = repository.getPetsForUser(currentUserId).first()
                val filteredAndSorted = applyFiltersAndSort(petsList)
                _pets.value = filteredAndSorted
            } catch (e: Exception) {
                _error.value = e.message
            }
        }
    }

    fun sortByName() {
        Timber.d("Sorting by name")
        _sortOrder.value = "name"
        refreshPets()
    }

    fun sortByDate() {
        Timber.d("Sorting by date")
        _sortOrder.value = "date"
        refreshPets()
    }

    fun clearSort() {
        Timber.d("Clearing sort")
        _sortOrder.value = "none"
        refreshPets()
    }

    fun clearError() {
        _error.value = null
    }

    // For accessing all pets (for the filter spinner)
    fun getAllPets(): LiveData<List<Pet>> = _petsLiveData

    // For vaccination history
    fun getVaccineHistory(): LiveData<List<VaccineHistoryItem>> {
        return repository.getVaccineHistory().asLiveData()
    }

    fun getVaccineHistoryForPet(petId: Long): LiveData<List<VaccineHistoryItem>> {
        return repository.getVaccineHistoryForPet(petId).asLiveData()
    }

    // Getter methods for Java interop
    fun getPetsLiveData(): LiveData<List<Pet>> = _petsLiveData
    fun getCurrentPetLiveData(): LiveData<Pet?> = _currentPetLiveData
    fun getVaccinesLiveData(): LiveData<List<Vaccine>> = _vaccinesLiveData
    fun getIsLoadingLiveData(): LiveData<Boolean> = _isLoadingLiveData
    fun getErrorLiveData(): LiveData<String?> = _errorLiveData
} 